MaxText/layers.py

```diff
@@ -704,8 +705,26 @@ def attend(self, query: Array) -> Array:
       in NLP models.
     """
     dtype = self.attend_dtype if self.attend_dtype is not None else self.dtype
-    return jnp.dot(query, jnp.asarray(self.embedding, dtype).T)
-
+    if not self.config.int8_training:
+      #return jnp.dot(query, jnp.asarray(self.embedding, dtype).T)
+      return lax.dot_general(query, jnp.asarray(self.embedding, dtype).T, (((query.ndim - 1,),(0,)),((),())))
+    else:
+      aqt_cfg = maxtext_sweeps.sweep1(
+        self.config.fwd_int8_qk,
+        self.config.dlhs_int8_qk,
+        self.config.drhs_int8_qk,
+        use_dummy_static_bound=self.config.aqt_use_dummy_static_bound,
+        rng_type=self.config.aqt_rng_type,
+        use_fwd_quant=self.config.aqt_use_fwd_quant,
+      )
+      aqt_dot_general = aqt.make_dot_general(aqt_cfg)
+      aqt_key = self.make_rng('aqt')
+      context = aqt.Context(key=aqt_key, train_step=None)
+      aqt_dot_general = functools.partial(aqt_dot_general, context=context)
+      # !!! AQT requires LHS dtype equals RHS dtype
+      # return aqt_dot_general(query, jnp.asarray(self.embedding).T, (((query.ndim - 1,),(0,)),((),())))
+      # The below fails with jnp.bfloat16
+      return aqt_dot_general(jnp.asarray(query, jnp.float32), jnp.asarray(self.embedding, jnp.float32).T, (((query.ndim - 1,),(0,)),((),())))
```

Comment:
We need to find common type, if both are f32, we are good, if bothr are bf16 we are good.
If they are bf16+f32 we need to cast both to f32

Reply:
Done.